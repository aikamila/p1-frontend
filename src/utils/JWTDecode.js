import jwt_decode from 'jwt-decode'

export const decodeUser = (data) => {
    return jwt_decode(data.access).user_id
}

