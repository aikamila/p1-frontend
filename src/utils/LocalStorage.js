export const getItemAndCheckExpiry = (key) => {
    const itemStr = localStorage.getItem(key)
    if (!itemStr) {
        return null
    }
    const item = JSON.parse(itemStr)
    const now = new Date()
    if (now.getTime() > item.expiry) {
        localStorage.removeItem(key)
        return null
    }
    return item.value
}


export const setItemwith8MonthExpiry = (key, value) => {
    const now = new Date()
    const item = {
        value: value,
        expiry: Date.now() + 1000 * 60 * 60 * 24 * 30 * 8,
    }
    localStorage.setItem(key, JSON.stringify(item))
}
