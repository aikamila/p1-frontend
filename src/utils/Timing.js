export const calculateTime = (hours, minutes, seconds) =>{
    return hours * 60 * 60 * 1000 + minutes * 60 * 1000 + seconds * 1000
}