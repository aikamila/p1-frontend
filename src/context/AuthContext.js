import { createContext, useState, useEffect, useCallback } from "react";
import { useHistory, useLocation } from "react-router-dom";
import { getItemAndCheckExpiry, setItemwith8MonthExpiry } from "../utils/LocalStorage";
import { calculateTime } from "../utils/Timing";
import { decodeUser } from "../utils/JWTDecode";

const AuthContext = createContext()

export default AuthContext;

export const AuthProvider = ({children}) => {

    const history = useHistory()

    let location = useLocation()

    const [authTokens, setAuthTokens] = useState(null)
    const [userId, setUserId] = useState(null)

    const [loginEmail, setLoginEmail] = useState('')
    const [loginPassword, setLoginPassword] = useState('')

    // const [pwResetEmail, setPwResetEmail] = useState('')
    // const [pwResetPassword, setPwResetPassword] = useState('')
    // const [pwResetPasswordSecond, setPwResetPasswordSecond] = useState('')

    const [loginError, setLoginError] = useState(null) // invalid credentials
 
    const [authServerError, setAuthServerError] = useState(null) // server error while logging in
    const [authSignUpServerError, setAuthSignUpServerError] = useState(null) // server error while signing up
    const [unexpectedLogoutError, setUnexpectedLogoutError] = useState('Something went wrong and you were logged out.')
    const [initialVerificationError, setInitialVerificationError] = useState(null)
    const [initialVerificationServerError, setInitialVerificationServerError] = useState(null)
    const [initialVerificationSuccessful, setInitialVerificationSuccessful] = useState(true)

    const [loginButtonDisabled, setLoginButtonDisabled] = useState(false)

    const BACKEND_DOMAIN = 'https://arcane-spire-03245.herokuapp.com/'
    const TOKENS_OBTAIN_URL = BACKEND_DOMAIN + 'api/user/token/'
    const TOKENS_REFRESH_URL = BACKEND_DOMAIN + 'api/user/token/refresh/'
    const TOKENS_BLACKLIST_URL = BACKEND_DOMAIN + 'api/user/token/blacklist/'
    const INITIAL_EMAIL_VERIFICATION = BACKEND_DOMAIN + 'api/user/token/verification/'

    const login = useCallback(async (e) => {
        e.preventDefault();
        setLoginButtonDisabled(true)
        try{
            const response = await fetch(TOKENS_OBTAIN_URL ,{
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({'email': loginEmail, 'password': loginPassword})
            })
            const data = await response.json()
            if (response.status === 401) {
                throw Error('Invalid credentials')
            }else if (response.status === 200){
                setUserId(decodeUser(data))
                setItemwith8MonthExpiry('authTokens', data)
                setAuthTokens(data)
                setLoginPassword('')
                history.push('/home')
            }else if(loginEmail.length === 0 || loginPassword.length === 0){
                throw Error('Invalid credentials')
            }else{
                throw Error('Something else happened')
            }
        }catch (err){
            switch (err.message){
                case 'Invalid credentials':
                    setLoginError(err.message)
                    setLoginPassword('')
                    break;
                default:
                    setAuthServerError('Upps... unexpected problem with the server occured. Try to refresh the site and log in again :)')
            }
        }
        setLoginButtonDisabled(false)
    }, 
    [TOKENS_OBTAIN_URL, loginEmail, loginPassword, history]
    )

    const logout = useCallback(async () => {
        try{
            await fetch(TOKENS_BLACKLIST_URL,{
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({'refresh': authTokens.refresh})
            })}catch(err){
                // doing nothing if the blacklisting of tokens goes wrong on the backend
            }
        const tokens = localStorage.getItem('authTokens')
        if (tokens){
            localStorage.removeItem('authTokens')
        }
        setAuthTokens(null)
        setUserId(null)
        history.push('/')
    }, [TOKENS_BLACKLIST_URL])

    const updateToken = useCallback(async () => {
        try{
            const tokens = getItemAndCheckExpiry('authTokens')
            // we get them from the local storage bc when the user comes back, they don't have the hook set (so we would send null :( )
            const response = await fetch(TOKENS_REFRESH_URL ,{
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({'refresh': tokens.refresh})
            })
            const data = await response.json()
            if (response.status !== 200) {
                throw Error('Invalid credentials')
            }else{
                setItemwith8MonthExpiry('authTokens', data)
                setAuthTokens(data)
                setUserId(decodeUser(data))
            }
        }catch(err){
            logout(); 
            setUnexpectedLogoutError('Something went wrong and you were logged out.')
            setTimeout(() => {
                setUnexpectedLogoutError(null)
            }, 5000)
        }
    }, [TOKENS_REFRESH_URL, logout]
    )

    const verifyEmail = async (id, token) => {
        try{
            const response = await fetch(INITIAL_EMAIL_VERIFICATION, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({'id': id, 'token': token})
                })
            if(response.status === 401){
                throw Error('Invalid link')
            }else if(response.status !== 200){
                throw Error("Something else happened")
            }
            const data = await response.json()
            if(response.status===200){
                await logout(); // if there is a user that is logged in, their token should be blacklisted
                setAuthTokens(data);
                setUserId(decodeUser(data))
                setItemwith8MonthExpiry('authTokens', data)
                setInitialVerificationSuccessful(true)
                setTimeout(() => {
                    setInitialVerificationSuccessful(false)
                }, 5000)
                history.push('/home')
            }
            }catch(err){
                if(err.message === 'Invalid link'){
                    setInitialVerificationError('The link is invalid. If you have already used it, just log into your account.')
                }else(
                    setInitialVerificationServerError('Something went wrong. Please use the link once again.')
                )
                history.push("/verification/unsuccessful")
            }
        
    }

    useEffect(() => {
        const fourMinutes = calculateTime(0, 4, 0)
        let interval = setInterval(() => {
            if (authTokens){
                updateToken()
            }
        }, fourMinutes)
        return () => clearInterval(interval)
    }, [authTokens, updateToken])

    useEffect(() => {
        const updateTokenAtTheBeginning = async () => {
            const tokens = getItemAndCheckExpiry('authTokens')
            if(tokens !== null){
                await updateToken();
            }
        }   
        updateTokenAtTheBeginning()
    }, [updateToken])

    useEffect(() => {
        setLoginPassword('')
        setLoginError(null)
        setAuthSignUpServerError(null)
        setAuthServerError(null)
    }, [location])

    let contextData = {
        login: login,
        loginEmail: loginEmail,
        setLoginEmail: setLoginEmail,
        loginPassword: loginPassword,
        setLoginPassword: setLoginPassword,
        loginError: loginError,
        authTokens: authTokens, 
        logout: logout,
        authServerError: authServerError,
        unexpectedLogoutError: unexpectedLogoutError,
        verifyEmail: verifyEmail,
        initialVerificationError: initialVerificationError,
        initialVerificationServerError: initialVerificationServerError,
        initialVerificationSuccessful: initialVerificationSuccessful,
        setAuthSignUpServerError: setAuthSignUpServerError,
        authSignUpServerError: authSignUpServerError,
        BACKEND_DOMAIN: BACKEND_DOMAIN,
        setLoginError: setLoginError,
        userId: userId,
        updateToken: updateToken,
        loginButtonDisabled: loginButtonDisabled,
    }

    return(
        <AuthContext.Provider value={contextData}>
            {children}
        </AuthContext.Provider>
    )}