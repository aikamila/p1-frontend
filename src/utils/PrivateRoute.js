import { Route, Redirect } from "react-router-dom";
import React, {useContext} from 'react';
import AuthContext from "../context/AuthContext";

const PrivateRoute = ({children, ...rest}) => {
    let {authTokens} = useContext(AuthContext)
    return (
        <Route {...rest}>{!authTokens ? <Redirect to='/auth'/> : children}</Route>
    )
}

export default PrivateRoute
