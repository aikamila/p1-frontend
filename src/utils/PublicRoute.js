import { Route, Redirect } from "react-router-dom";
import React, {useContext} from 'react';
import AuthContext from "../context/AuthContext";


const PublicRoute = ({children, ...rest}) => {
    let {authTokens} = useContext(AuthContext)
    return (
        <Route {...rest}>{authTokens ? <Redirect to='/home'/> : children}</Route>
    )
}

export default PublicRoute