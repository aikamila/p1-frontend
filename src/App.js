import logo from './logo.svg';
import './App.css';
import { BrowserRouter as Router, Route, Switch} from 'react-router-dom';
import StartPage from './pages/StartPage';
import PrivateRoute from './utils/PrivateRoute';
import PublicRoute from './utils/PublicRoute';

import AuthPage from './pages/AuthPage';
import HomePage from './pages/HomePage';
import InitialVerificationPage from './pages/InitialVerificationPage';
import UnsuccessfulInitialVerificationPage from './pages/UnsuccessfulInitialVerificationPage';

function App() {
  return (
     <div className="App">

          <PublicRoute path="/" exact>
            <StartPage/>
          </PublicRoute>
          <PublicRoute path="/auth">
            <AuthPage/>
          </PublicRoute>
          <PrivateRoute path="/home"> 
            <HomePage/>
          </PrivateRoute>
            <Switch>
              <Route path="/verification/unsuccessful" component={UnsuccessfulInitialVerificationPage}/>
              <Route path="/verification/:id/:token" component={InitialVerificationPage} />
            </Switch>

     </div>

  );
}

export default App;
