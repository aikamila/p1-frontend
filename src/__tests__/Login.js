import { render, screen, waitFor } from '@testing-library/react'
import App from '../App'
import { AuthProvider } from '../context/AuthContext'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import * as storageUtils from '../utils/LocalStorage'
import userEvent from '@testing-library/user-event'
import * as timingUtils from "../utils/Timing"
import renderer from 'react-test-renderer'
import { server } from '../mocks/server'
import { rest } from 'msw'


afterEach(() => {
  localStorage.removeItem("authTokens")
})

jest.setTimeout(6000)

function renderLoginPage(history) {
  history.push('/auth')
  return render(
    <Router history = {history}>
      <AuthProvider>
        <App />
      </AuthProvider>
    </Router>
  );
}

test("all fields work properly + behavior when invalid email or password", async () => {
  const history=createMemoryHistory()
  renderLoginPage(history)
  const email = screen.getByRole('textbox', {name: /email/i})
  const password = screen.getByRole('textbox', {name: /password/i})
  expect(email).toBeInTheDocument()
  expect(password).toBeInTheDocument()
  expect(screen.getByPlaceholderText(/email/i)).toBeInTheDocument()
  expect(screen.getByPlaceholderText(/password/i)).toBeInTheDocument()
  userEvent.type(email, 'invalid@mail.com')
  expect(email).toHaveValue('invalid@mail.com')
  userEvent.type(password, 'Password')
  expect(password).toHaveValue('Password')
  const submitButton = screen.getByRole('button', {name: /log in/i})
  expect(submitButton).toBeInTheDocument()
  expect(submitButton).not.toBeDisabled()
  userEvent.click(submitButton)
  expect(submitButton).toBeDisabled()
  const invalidCredentials = await screen.findByText(/invalid credentials/i)
  expect(submitButton).not.toBeDisabled()
  let alerts = screen.queryAllByRole("alert")
  expect(alerts).toHaveLength(1)
  expect(invalidCredentials).toBeInTheDocument()
  expect(screen.getByRole('textbox', {name: /email/i})).toHaveStyle('border: 1px solid red')
  expect(screen.getByRole('textbox', {name: /password/i})).toHaveStyle('border: 1px solid red')
  expect(password).toHaveValue('')
  userEvent.type(password, "D")
  alerts = screen.queryAllByRole('alert')
  expect(alerts).toHaveLength(0)
  expect(screen.queryByText(/invalid credentials/i)).not.toBeInTheDocument()
  expect(screen.getByRole('textbox', {name: /email/i})).not.toHaveStyle('border: 1px solid red')
  expect(screen.getByRole('textbox', {name: /password/i})).not.toHaveStyle('border: 1px solid red')
  userEvent.click(submitButton)
  alerts = await screen.findAllByRole("alert")
  expect(alerts).toHaveLength(1)
  expect(await screen.findByText(/invalid credentials/i)).toBeInTheDocument()
  userEvent.type(email, "D")
  expect(screen.queryByText(/invalid credentials/i)).not.toBeInTheDocument()
  expect(screen.getByRole('textbox', {name: /email/i})).not.toHaveStyle('border: 1px solid red')
  expect(screen.getByRole('textbox', {name: /password/i})).not.toHaveStyle('border: 1px solid red')
  alerts = screen.queryAllByRole('alert')
  expect(alerts).toHaveLength(0)
  expect(password.value).toBe('')
  expect(localStorage.getItem("authTokens")).toBeNull()
})

test("password value disappears when the user changes the url - email doesn't", () => {
  const history=createMemoryHistory()
  renderLoginPage(history)
  const email = screen.getByRole('textbox', {name: /email/i})
  const password = screen.getByRole('textbox', {name: /password/i})
  userEvent.type(email, 'invalid@mail.com')
  userEvent.type(password, 'Password')
  history.push('/')
  history.push('/auth')
  expect(screen.getByRole('textbox', {name: /email/i})).toHaveValue('invalid@mail.com')
  expect(screen.getByRole('textbox', {name: /password/i})).toHaveValue('')
})


describe("all links work properly", () => {
  test("sigup link works properly", () => {
    const history=createMemoryHistory()
    renderLoginPage(history)
    const signupLink = screen.queryByRole("link", {name: /sign up instead/i})
    expect(signupLink).toBeInTheDocument()
    expect(signupLink).toHaveAttribute("href", "/auth/signup")
  })
  test("start page link works properly", () => {
    const history=createMemoryHistory()
    renderLoginPage(history)
    const startPageLink = screen.queryByRole("link", {name: /go back to the homepage/i})
    expect(startPageLink).toBeInTheDocument()
    expect(startPageLink).toHaveAttribute("href", "/")
  })
})

test("invalid credentials alert disappears after changing the url", async () => {
  const history=createMemoryHistory()
  renderLoginPage(history)
  const email = screen.getByRole('textbox', {name: /email/i})
  const password = screen.getByRole('textbox', {name: /password/i})
  userEvent.type(email, 'invalid@mail.com')
  userEvent.type(password, 'Password')
  const submitButton = screen.getByRole('button', {name: /log in/i})
  userEvent.click(submitButton)
  const invalidCredentials = await screen.findByText(/invalid credentials/i)
  expect(invalidCredentials).toBeInTheDocument()  
  expect(screen.getByRole('textbox', {name: /email/i})).toHaveStyle('border: 1px solid red')
  expect(screen.getByRole('textbox', {name: /password/i})).toHaveStyle('border: 1px solid red')
  history.push("/")
  history.push("/auth")
  expect(screen.queryByText(/invalid credentials/i)).not.toBeInTheDocument()
  expect(screen.getByRole('textbox', {name: /email/i})).not.toHaveStyle('border: 1px solid red')
  expect(screen.getByRole('textbox', {name: /password/i})).not.toHaveStyle('border: 1px solid red')
})

test("server error alert appears and then disappears after changing the url", async () => {
    const history = createMemoryHistory()
    renderLoginPage(history)
    const password = screen.getByRole('textbox', {name: /password/i})
    userEvent.type(password, "StrongPassword")
    const login = screen.getByRole('textbox', {name: /email/i})
    userEvent.type(login, 'servererror@mail.com')
    userEvent.click(screen.getByRole('button', {name: 'Log in'}))
    expect(await screen.findByText(/unexpected problem with the server/i)).toBeInTheDocument();
    expect(screen.queryByRole('textbox', {name: /password/i})).not.toBeInTheDocument()
    expect(screen.queryByRole('textbox', {name: /email/i})).not.toBeInTheDocument()
    expect(screen.queryByRole('button', {name: /Log in/i})).not.toBeInTheDocument();
    expect(screen.queryByRole('link', {name: /to the homepage/i})).not.toBeInTheDocument()
    expect(screen.queryByText(/sign up instead/i)).not.toBeInTheDocument()
    history.push('/')
    history.push('/auth')
    expect(screen.queryByText(/unexpected problem with the server/i)).not.toBeInTheDocument()
    expect(screen.queryByRole('textbox', {name: /password/i})).toBeInTheDocument()
    expect(screen.queryByRole('textbox', {name: /email/i})).toBeInTheDocument()
    expect(screen.queryByRole('button', {name: /Log in/i})).toBeInTheDocument();
    expect(screen.queryByRole('link', {name: /to the homepage/i})).toBeInTheDocument()
    expect(screen.queryByText(/sign up instead/i)).toBeInTheDocument()
    expect(localStorage.getItem("authTokens")).toBeNull()
})

test("server error alert appears when an unexpected response staus is received and the error disappears after changng the url", async () => {
  const history = createMemoryHistory()
  renderLoginPage(history)
  const password = screen.getByRole('textbox', {name: /password/i})
  userEvent.type(password, "StrongPassword")
  const login = screen.getByRole('textbox', {name: /email/i})
  userEvent.type(login, "fourohfour@mail.com")
  userEvent.click(screen.getByRole('button', {name: 'Log in'}))
  expect(await screen.findByText(/unexpected problem with the server/i)).toBeInTheDocument();
  expect(screen.queryByRole('textbox', {name: /password/i})).not.toBeInTheDocument()
  expect(screen.queryByRole('textbox', {name: /email/i})).not.toBeInTheDocument()
  expect(screen.queryByRole('button', {name: /Log in/i})).not.toBeInTheDocument();
  expect(screen.queryByRole('link', {name: /to the homepage/i})).not.toBeInTheDocument()
  expect(screen.queryByText(/sign up instead/i)).not.toBeInTheDocument()
  history.push('/')
  history.push('/auth')
  expect(screen.queryByText(/unexpected problem with the server/i)).not.toBeInTheDocument()
  expect(screen.queryByRole('textbox', {name: /password/i})).toBeInTheDocument()
  expect(screen.queryByRole('textbox', {name: /email/i})).toBeInTheDocument()
  expect(screen.queryByRole('button', {name: /Log in/i})).toBeInTheDocument();
  expect(screen.queryByRole('link', {name: /to the homepage/i})).toBeInTheDocument()
  expect(screen.queryByText(/sign up instead/i)).toBeInTheDocument()
  expect(localStorage.getItem("authTokens")).toBeNull()
})

test("user logs in and everything works as expected", async () => {
  const history=createMemoryHistory()
  renderLoginPage(history)
  const calculateTimeMock = jest.spyOn(timingUtils, 'calculateTime')
  expect(calculateTimeMock(0,4,0)).toEqual(240000)
  // mock implementation that reduces the time in which tokens are updated x1000
  calculateTimeMock.mockImplementation((hours, minutes, seconds) => {return hours * 60 * 60 + minutes * 60 + seconds})
  const setItemWith8MonthExpiryMock = jest.spyOn(storageUtils, "setItemwith8MonthExpiry")
  const email = screen.getByRole('textbox', {name: /email/i})
  const password = screen.getByRole('textbox', {name: /password/i})
  userEvent.type(email, 'valid@mail.com')
  userEvent.type(password, 'Password')
  const submitButton = screen.getByRole('button', {name: /log in/i})
  userEvent.click(submitButton)
  await waitFor(() => expect(history.location.pathname).toBe('/home'))
  await waitFor(() => expect(localStorage.getItem('authTokens')).toContain("valid-access"))
  await waitFor(() => expect(localStorage.getItem('authTokens')).toContain("new-access"))
  await waitFor(() => expect(localStorage.getItem('authTokens')).toContain("new-new-access"))
  expect(setItemWith8MonthExpiryMock).toHaveBeenCalledTimes(3) // checking if tokens are set for 8 months
  expect(calculateTimeMock).toHaveBeenLastCalledWith(0, 4, 0)
  expect(history.location.pathname).toBe('/home')
  calculateTimeMock.mockRestore()
  setItemWith8MonthExpiryMock.mockRestore()
})

test("user logs in but gets logged out - the refresh token is invalid", async () => {
  const history=createMemoryHistory()
  renderLoginPage(history)
  const calculateTimeMock = jest.spyOn(timingUtils, 'calculateTime')
  calculateTimeMock.mockImplementation((hours, minutes, seconds) => {return hours * 60 * 60 + minutes * 60 + seconds})
  const email = screen.getByRole('textbox', {name: /email/i})
  const password = screen.getByRole('textbox', {name: /password/i})
  userEvent.type(email, 'invalidtoken@mail.com')
  userEvent.type(password, 'Password')
  const submitButton = screen.getByRole('button', {name: /log in/i})
  userEvent.click(submitButton)
  await waitFor(() => expect(history.location.pathname).toBe('/home'))
  await waitFor(() => expect(localStorage.getItem('authTokens')).toContain("invalid-refresh")) 
  await waitFor(() => expect(history.location.pathname).toBe('/'))
  expect(screen.getByText(/Something went wrong and you were logged out/i)).toBeInTheDocument()
  expect(localStorage.getItem('authTokens')).toBeNull()
  expect(calculateTimeMock).toHaveBeenLastCalledWith(0, 4, 0)
  await waitFor(() => expect(screen.queryByText(/Something went wrong and you were logged out/i)).not.toBeInTheDocument(), {timeout: 5050})
  calculateTimeMock.mockRestore()
})

test("user has been logged in previously - tokens are not refreshed because of a server error", async () => {
  storageUtils.setItemwith8MonthExpiry("authTokens", {access: "valid-access", refresh: "servererror"})
  const history=createMemoryHistory()
  renderLoginPage(history)
  expect(localStorage.getItem('authTokens')).toContain("servererror")
  await waitFor(() => expect(history.location.pathname).toBe("/"))
  expect(screen.getByText(/Something went wrong and you were logged out/i)).toBeInTheDocument()
  expect(localStorage.getItem('authTokens')).toBeNull()
  await waitFor(() => expect(screen.queryByText(/Something went wrong and you were logged out/i)).not.toBeInTheDocument(), {timeout: 5050})
})

test("login form is rendered correctly", () => {
  const history = createMemoryHistory();
  history.push("/auth")
  const page = renderer
    .create(
      <Router history = {history}>
        <AuthProvider>
          <App />
        </AuthProvider>
      </Router>
    )
    .toJSON()
    expect(page).toMatchSnapshot()
})

test("email is saved in the form and is still visible when the user logs out", async () => {
  const history=createMemoryHistory()
  renderLoginPage(history)
  const email = screen.getByRole('textbox', {name: /email/i})
  const password = screen.getByRole('textbox', {name: /password/i})
  userEvent.type(email, 'valid@mail.com')
  userEvent.type(password, 'Password')
  const submitButton = screen.getByRole('button', {name: /log in/i})
  userEvent.click(submitButton)
  const logoutButton = await screen.findByRole("button", {name: /log out/i})
  userEvent.click(logoutButton)
  await waitFor(() => expect(history.location.pathname).toBe("/"))
  history.push("/auth")
  expect(screen.getByRole("textbox", {name: /email/i})).toHaveValue("valid@mail.com")
})

test("user has been logged in previously - everything works as expected", async() => {
  const calculateTimeMock = jest.spyOn(timingUtils, 'calculateTime')
  storageUtils.setItemwith8MonthExpiry("authTokens", {access: "valid-access", refresh: "valid-refresh"})
  calculateTimeMock.mockImplementation((hours, minutes, seconds) => {return hours * 60 * 60 + minutes * 60 + seconds})
  const setItemWith8MonthExpiryMock = jest.spyOn(storageUtils, "setItemwith8MonthExpiry")
  const history=createMemoryHistory()
  renderLoginPage(history)
  expect(localStorage.getItem('authTokens')).toContain("valid-refresh")
  await waitFor(() => expect(history.location.pathname).toBe("/home"))
  await waitFor(() => expect(localStorage.getItem('authTokens')).toContain("new-refresh"))
  await waitFor(() => expect(localStorage.getItem('authTokens')).toContain("new-new-refresh"))
  expect(history.location.pathname).toBe("/home")
  expect(setItemWith8MonthExpiryMock).toHaveBeenCalledTimes(2) // checking if tokens are set for 8 months
  calculateTimeMock.mockRestore()
  setItemWith8MonthExpiryMock.mockRestore()
})

test("auth tokens disappear from local storage after 8 months", () => {
  storageUtils.setItemwith8MonthExpiry("authTokens", {access: "valid", refresh: "valid"})
  jest.useFakeTimers()
  jest.advanceTimersByTime(1000 * 60 * 60 * 24 * 31 * 7)
  expect(storageUtils.getItemAndCheckExpiry("authTokens")).not.toBeNull()
  jest.advanceTimersByTime(1000 * 60 * 60 * 24 * 31 * 1)
  expect(storageUtils.getItemAndCheckExpiry("authTokens")).toBeNull()
  jest.useRealTimers()
})

test("user logs in but gets logged out - server responds with an unexpected response status (while refreshing tokens)", async () => {
  const history=createMemoryHistory()
  renderLoginPage(history)
  const calculateTimeMock = jest.spyOn(timingUtils, 'calculateTime')
  calculateTimeMock.mockImplementation((hours, minutes, seconds) => {return hours * 60 * 60 + minutes * 60 + seconds})
  const email = screen.getByRole('textbox', {name: /email/i})
  const password = screen.getByRole('textbox', {name: /password/i})
  userEvent.type(email, '404token@mail.com')
  userEvent.type(password, 'Password')
  const submitButton = screen.getByRole('button', {name: /log in/i})
  userEvent.click(submitButton)
  await waitFor(() => expect(history.location.pathname).toBe('/home'))
  await waitFor(() => expect(localStorage.getItem('authTokens')).toContain("404")) 
  await waitFor(() => expect(localStorage.getItem('authTokens')).toContain("valid-access")) 
  await waitFor(() => expect(history.location.pathname).toBe('/'))
  expect(screen.getByText(/Something went wrong and you were logged out/i)).toBeInTheDocument()
  expect(localStorage.getItem('authTokens')).toBeNull()
  expect(calculateTimeMock).toHaveBeenLastCalledWith(0, 4, 0)
  await waitFor(() => expect(screen.queryByText(/Something went wrong and you were logged out/i)).not.toBeInTheDocument(), {timeout: 5050})
  calculateTimeMock.mockRestore()
})

test("bug fix -- if one or more of the fields are empty and the server response is not 401, the user should see 'invalid credentials' instead of server error", async () => {
  server.use(
    rest.post('https://arcane-spire-03245.herokuapp.com/api/user/token/', (req, res, ctx) => {
      return res.once(
        ctx.status(400),
        ctx.json({})
      )
    })
  )
  server.use(
    rest.post('https://arcane-spire-03245.herokuapp.com/api/user/token/', (req, res, ctx) => {
      return res.once(
        ctx.status(404),
        ctx.json({})
      )
    })
  )
  const history = createMemoryHistory()
  renderLoginPage(history)
  userEvent.type(screen.getByRole("textbox", {name: /email/i}), "ee")
  userEvent.click(screen.getByRole("button", {name: /log in/i}))
  expect(await screen.findByText(/invalid credentials/i)).toBeInTheDocument()
  userEvent.type(screen.getByRole("textbox", {name: /email/i}), "{backspace}{backspace}")
  screen.debug()
  userEvent.type(screen.getByRole("textbox", {name: /password/i}), "ee")
  userEvent.click(screen.getByRole("button", {name: /log in/i}))
  expect(await screen.findByText(/invalid credentials/i)).toBeInTheDocument()
})