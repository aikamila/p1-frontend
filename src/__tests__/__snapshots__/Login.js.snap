// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login form is rendered correctly 1`] = `
<div
  className="App"
>
  <header
    className="auth-header"
    role="banner"
  >
    <img
      alt="shangify brand logo"
      className="auth-header__img1"
      src="shangify_logo.png"
    />
    <img
      alt=""
      className="auth-header__img2"
      src="shangify_brand_name.png"
    />
    <nav
      aria-label="main navigation"
      className="auth-header__nav"
      role="navigation"
    >
      <a
        aria-label="Go back to the homepage"
        className="auth-header__homepage_link"
        href="/"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
          />
        </svg>
      </a>
    </nav>
  </header>
  <main
    className="log-in-form__background"
  >
    <div
      className="log-in-form__window"
    >
      <div
        className="log-in-form__el cta"
      >
        <p>
          Welcome again!
          <br />
          Log in and share your creative ideas with the world!
        </p>
      </div>
      <div
        className="log-in-form__el email"
      >
        <input
          aria-label="email"
          className="log-in-form__input"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          style={Object {}}
          type="text"
          value=""
        />
      </div>
      <div
        className="log-in-form__el password"
      >
        <input
          aria-label="password"
          className="log-in-form__input"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          role="textbox"
          style={Object {}}
          type="password"
          value=""
        />
      </div>
      <button
        className="log-in-form__button"
        disabled={false}
        onClick={[Function]}
      >
        Log in
      </button>
      <p
        className="log-in-form__signup"
      >
        I wanna 
        <a
          className="log-in-form__signup_link"
          href="/auth/signup"
          onClick={[Function]}
        >
          sign up instead
        </a>
      </p>
    </div>
  </main>
</div>
`;
